File: core/providers/factory.py
"""按 settings.yaml 的 provider 配置构造 LLMProvider 实例。

新增供应商 = 在 core/providers/ 下新增实现类 + 在此注册一行。
"""
from __future__ import annotations

from core.providers.anthropic_compat import AnthropicCompatProvider
from core.providers.base import LLMProvider
from core.providers.gemini import GeminiProvider

DEFAULT_MODELS = {
    "gemini": "gemini-flash-latest",
}


def create_provider(name: str, opts: dict | None = None) -> LLMProvider:
    opts = opts or {}
    if name == "gemini":
        return GeminiProvider(model=opts.get("model", DEFAULT_MODELS["gemini"]))
    if name == "anthropic-compat":
        return AnthropicCompatProvider(
            model=opts["model"],
            base_url=opts.get("base_url"),
            api_key_env=opts.get("api_key_env", "ANTHROPIC_API_KEY"),
            max_tokens=opts.get("max_tokens", 16384),
        )
    # openai-compat 按 TASKS T4.2 在 Phase 4 实现
    raise ValueError(f"未知 provider: {name!r}")


def resolve_provider(provider_cfg: dict) -> LLMProvider:
    """按 primary → fallback 顺序返回第一个可构造的 provider（缺 Key 自动跳过）。"""
    problems: list[str] = []
    for name in (provider_cfg.get("primary"), provider_cfg.get("fallback")):
        if not name:
            continue
        try:
            return create_provider(name, provider_cfg.get(name) or {})
        except (RuntimeError, ValueError) as e:
            problems.append(f"{name}: {e}")
    raise RuntimeError(
        "无可用 LLM 供应商（请在 .env / 环境变量配置对应 Key）：\n  "
        + "\n  ".join(problems)
    )

File: src/providers/factory.ts
/** 供应商组装栈：RateLimitedProvider( FallbackProvider( primary, fallback ) )——
 *  令牌桶主动限速在最外；主供应商退避穷尽后切换备用（即使无备用也获得重试）。 */
import type { ZodType } from "zod";

import { withBackoff, TokenBucketLimiter } from "../rateLimiter.ts";
import { PiAiProvider } from "./piAiProvider.ts";
import type { ExtractionResult, LlmProvider } from "./base.ts";
import { TransientProviderError } from "./base.ts";

export class FallbackProvider implements LlmProvider {
  readonly primary: LlmProvider;
  readonly fallback: LlmProvider | null;
  private readonly maxRetries: number;
  private readonly onDegrade?: (from: string, to: string) => void;
  private readonly sleep?: (seconds: number) => Promise<void>;

  constructor(
    primary: LlmProvider,
    fallback: LlmProvider | null = null,
    maxRetries = 5,
    onDegrade?: (from: string, to: string) => void,
    sleep?: (seconds: number) => Promise<void>,
  ) {
    this.primary = primary;
    this.fallback = fallback;
    this.maxRetries = maxRetries;
    this.onDegrade = onDegrade;
    this.sleep = sleep;
  }

  get name(): string {
    return this.fallback ? `${this.primary.name}|${this.fallback.name}` : this.primary.name;
  }

  private retry<T>(fn: () => Promise<T>): Promise<T> {
    return withBackoff(fn, {
      maxRetries: this.maxRetries,
      sleep: this.sleep,
      isTransient: (e) => e instanceof TransientProviderError,
    });
  }

  async extract<T>(
    content: string,
    schema: ZodType<T>,
    opts: { instruction?: string } = {},
  ): Promise<ExtractionResult<T>> {
    try {
      return await this.retry(() => this.primary.extract(content, schema, opts));
    } catch (e) {
      if (!(e instanceof TransientProviderError) || !this.fallback) throw e;
      this.onDegrade?.(this.primary.name, this.fallback.name);
      return await this.retry(() => this.fallback!.extract(content, schema, opts));
    }
  }
}

export class RateLimitedProvider implements LlmProvider {
  readonly inner: LlmProvider;
  private readonly limiter: TokenBucketLimiter | null;

  constructor(inner: LlmProvider, rpm?: number) {
    this.inner = inner;
    this.limiter = rpm ? new TokenBucketLimiter(rpm) : null;
  }

  get name(): string {
    return this.inner.name;
  }

  async extract<T>(
    content: string,
    schema: ZodType<T>,
    opts: { instruction?: string } = {},
  ): Promise<ExtractionResult<T>> {
    if (this.limiter) await this.limiter.acquire();
    return this.inner.extract(content, schema, opts);
  }
}

const DEFAULT_MODELS: Record<string, string> = {
  "minimax-cn": "MiniMax-M3",
  google: "gemini-flash-latest", // 以 pi-ai 目录当前可用模型为准
};

interface ProviderSettings {
  primary?: string;
  fallback?: string;
  [key: string]: unknown;
}

/** 按 settings.yaml 的 provider 配置组装栈；全部缺 Key 抛错。 */
export function createProviderStack(providerCfg: ProviderSettings): LlmProvider {
  const problems: string[] = [];
  const built: [string, LlmProvider][] = [];
  for (const name of [providerCfg.primary, providerCfg.fallback]) {
    if (!name) continue;
    const opts = (providerCfg[name] ?? {}) as { model?: string; rpm?: number };
    try {
      built.push([name, new PiAiProvider(name, opts.model ?? DEFAULT_MODELS[name] ?? "")]);
    } catch (e) {
      problems.push(`${name}: ${e}`);
    }
  }
  if (built.length === 0) {
    throw new Error(`无可用 LLM 供应商（检查 .env / 环境变量）：\n  ${problems.join("\n  ")}`);
  }
  const [effectiveName, effective] = built[0];
  const inner = new FallbackProvider(effective, built[1]?.[1] ?? null);
  const rpm = ((providerCfg[effectiveName] ?? {}) as { rpm?: number }).rpm;
  return new RateLimitedProvider(inner, rpm);
}
